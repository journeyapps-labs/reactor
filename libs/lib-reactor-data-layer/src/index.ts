export * from './PaginatedCollection';
export * from './Collection';
export * from './LifecycleCollection';
