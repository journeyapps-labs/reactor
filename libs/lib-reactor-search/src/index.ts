export * from './SearchResult';
export * from './SearchEngineInterface';
