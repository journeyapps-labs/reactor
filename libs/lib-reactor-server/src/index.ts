export * from './ReactorModule';
export * from './ReactorConfig';
export * from './basic-html';
export * from './reactor-html';
