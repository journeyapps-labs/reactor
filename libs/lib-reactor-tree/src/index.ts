export * from './TreeEntity';
export * from './TreeNode';
export * from './TreeException';
