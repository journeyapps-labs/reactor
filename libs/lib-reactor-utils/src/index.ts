export * from './clipboard';
export * from './color';
export * from './file-input';
export * from './string-manipulation';
export * from './ReactorPath';
export * from './Mutex';
export * from './EventEmitter';
export * from './Disposable';
export * from './Logger';
export * from './Validators';
export * from './retries';
export * from './async-throttle';
export * from './BaseObserver';
export * from './dates';
