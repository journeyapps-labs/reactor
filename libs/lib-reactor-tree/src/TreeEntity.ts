import { TreeNode } from './TreeNode';
import { BaseListener, BaseObserver } from './BaseObserver';

export enum PathEntryType {
  CHILD = 'child',
  ROOT = 'index',
  ATTR = 'attr'
}

export type PathEntry = [PathEntryType, string | number];

export type Path = PathEntry[];

export interface TreeEntityInterface {
  getPath(): Path;
  getPathAsString(): string;
  getRoot(): TreeEntity;
  getKey(): string;
  delete();
  getParent<T extends TreeNode>(): T;
}

export interface FlattenOptions {
  filter?: (entity: TreeEntity) => boolean;
}

export interface TreeEntityListener extends BaseListener {
  deleted: () => any;
}

export class TreeEntity<T extends TreeEntityListener = TreeEntityListener>
  extends BaseObserver<T>
  implements TreeEntityInterface
{
  protected parent: TreeNode;
  private _cache: string;

  constructor(private key: string) {
    super();
    this._cache = null;
    this.parent = null;
  }

  getKey() {
    return this.key;
  }

  delete() {
    this.iterateListeners((cb) => cb.deleted?.());
  }

  setParent(parent: TreeNode) {
    this.parent = parent;
    this._cache = null;
  }

  getPathAsString(): string {
    if (!this._cache) {
      this._cache = JSON.stringify(this.getPath());
    }
    return this._cache;
  }

  open() {
    if (this.parent) {
      this.parent.open();
    }
  }

  getParent<T extends TreeNode>(): T {
    return this.parent as T;
  }

  getPathEntryType() {
    return PathEntryType.ATTR;
  }

  getPath(): Path {
    let path = [[this.getPathEntryType(), this.key]] as Path;
    if (!this.parent) {
      return path;
    }
    return [...this.parent.getPath(), ...path];
  }

  flatten(options?: FlattenOptions): TreeEntity[] {
    if (options?.filter?.(this) === false) {
      return [];
    }
    return [this];
  }

  getRoot(): TreeEntity {
    if (this.parent) {
      return this.parent.getRoot();
    }
    return this;
  }
}
