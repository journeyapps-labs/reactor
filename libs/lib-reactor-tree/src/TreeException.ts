import { TreeEntity } from './TreeEntity';

export class TreeException extends Error {
  constructor(
    public tree: TreeEntity,
    error: string
  ) {
    super(error);
  }
}
