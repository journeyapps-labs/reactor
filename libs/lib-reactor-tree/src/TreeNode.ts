import { FlattenOptions, Path, PathEntry, PathEntryType, TreeEntity, TreeEntityListener } from './TreeEntity';
import { TreeException } from './TreeException';

export interface TreeSerialized {
  [key: string]: { collapsed: boolean };
}

export interface TreeNodeListener extends TreeEntityListener {
  collapsedChanged: (collapsed: boolean) => any;
  childAdded: (child: TreeEntity) => any;
  childRemoved: (child: TreeEntity) => any;
}

export class TreeNode<T extends TreeNodeListener = TreeNodeListener> extends TreeEntity<T> {
  protected _childrenSet: Set<TreeEntity>;
  protected _childrenMap: Map<string, TreeEntity>;
  collapsed: boolean;

  constructor(key: string) {
    super(key);
    this._childrenSet = new Set();
    this._childrenMap = new Map();
    this.collapsed = true;
  }

  getNodes<T extends TreeNode>(): T[] {
    return this.children.filter((c) => c instanceof TreeNode) as T[];
  }

  getChildPath(element: TreeEntity): PathEntry[] {
    if (!this._childrenSet.has(element)) {
      throw new TreeException(this, `Element with path: [${element.getPathAsString()}] does not exist in this node`);
    }

    if (element instanceof TreeNode) {
      return this.getPath().concat([[PathEntryType.CHILD, element.getKey()]]);
    }
    return this.getPath().concat([[PathEntryType.ATTR, element.getKey()]]);
  }

  get children() {
    return Array.from(this._childrenSet.values());
  }

  getPathEntryType() {
    return PathEntryType.CHILD;
  }

  open() {
    this.setCollapsed(false);
    super.open();
  }

  deserialize(payload: TreeSerialized) {
    for (let key in payload) {
      const node = this.fromPath(key);
      if (node && node instanceof TreeNode) {
        node.setCollapsed(payload[key].collapsed);
      }
    }
  }

  serialize(): TreeSerialized {
    const object: TreeSerialized = {};
    for (let ob of this.flatten()) {
      if (ob instanceof TreeNode) {
        object[ob.getPathAsString()] = {
          collapsed: ob.collapsed
        };
      }
    }
    return object;
  }

  setCollapsed(collapsed: boolean = true) {
    if (collapsed !== this.collapsed) {
      this.collapsed = collapsed;
      this.iterateListeners((cb) => cb.collapsedChanged?.(collapsed));
    }
  }

  openChildren(open: boolean = true) {
    for (let child of this._childrenSet) {
      if (child instanceof TreeNode) {
        child.openChildren(open);
      }
    }
    this.setCollapsed(!open);
  }

  toggle() {
    this.setCollapsed(!this.collapsed);
  }

  getChild<T extends TreeEntity>(key: string): T {
    return this._childrenMap.get(key) as T;
  }

  addChild(child: TreeEntity) {
    if (this._childrenMap.has(child.getKey())) {
      throw new TreeException(this, `Tree with key ${child.getKey()} already exists in this node`);
    }
    child.setParent(this);
    this._childrenSet.add(child);
    this._childrenMap.set(child.getKey(), child);
    const l1 = child.registerListener({
      deleted: () => {
        this._childrenSet.delete(child);
        this._childrenMap.delete(child.getKey());
        child.setParent(null);
        l1?.();
        this.iterateListeners((cb) => cb.childRemoved?.(child));
      }
    });
    this.iterateListeners((cb) => cb.childAdded?.(child));
  }

  clear() {
    this._childrenSet.forEach((c) => {
      c.delete();
    });
  }

  setChildren(children: TreeEntity[]) {
    this.clear();
    children.forEach((c) => {
      this.addChild(c);
    });
  }

  flatten(options?: FlattenOptions): TreeEntity[] {
    let arr: TreeEntity[] = [];

    if (options?.filter?.(this) !== false) {
      arr.push(this);
    }

    for (let child of this._childrenSet) {
      arr = arr.concat(child.flatten(options));
    }
    return arr;
  }

  fromPath(path: Path | string): TreeEntity {
    if (typeof path !== 'string') {
      path = JSON.stringify(path);
    }
    const flatten = this.flatten();
    for (let f of flatten) {
      if (f.getPathAsString() === path) {
        return f;
      }
    }
    return null;
  }
}
