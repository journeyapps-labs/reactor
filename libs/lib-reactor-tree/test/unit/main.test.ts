import { TreeEntity, TreeNode } from '../../src';
import { describe, it, expect } from 'vitest';

describe('Tree', () => {
  it('should test all tree things', async () => {
    const root = new TreeNode('root');
    const child1 = new TreeNode('child1');
    const child2 = new TreeNode('child2');
    const child3 = new TreeEntity('child3');

    root.addChild(child1);
    child1.addChild(child2);
    root.addChild(child3);

    // check parents
    expect(child1.getParent()).toBe(root);
    expect(root.flatten().length).toEqual(4);
    expect(root.getChild('child1')).toBe(child1);
    expect(child1.getChild('child2')).toBe(child2);
    await expect(() => {
      root.addChild(new TreeNode('child1'));
    }).toThrow();

    // deep fetch
    expect(root.fromPath(child1.getPath())).toBe(child1);
    expect(root.fromPath(child2.getPath())).toBe(child2);
    expect(root.fromPath(child3.getPath())).toBe(child3);

    // deleting
    child3.delete();
    expect(child3.getParent()).toEqual(null);
    expect(root.flatten().length).toEqual(3);
    root.addChild(child3);

    // serialization
    const serialized = root.serialize();
    expect(serialized).toMatchSnapshot();
    expect(root.collapsed).toEqual(true);
    expect(child1.collapsed).toEqual(true);
    expect(child2.collapsed).toEqual(true);

    const promises = [root, child1, child2].map((c: TreeNode) => {
      return new Promise((resolve) => {
        const remove = c.registerListener({
          collapsedChanged: (col) => {
            resolve(col);
            remove?.();
          }
        });
      });
    });

    root.openChildren(true);

    const res = await Promise.all(promises);
    expect(res[0]).toEqual(root.collapsed);
    expect(res[1]).toEqual(child1.collapsed);
    expect(res[2]).toEqual(child2.collapsed);

    expect(root.collapsed).toEqual(false);
    expect(child1.collapsed).toEqual(false);
    expect(child2.collapsed).toEqual(false);

    // add child 3 to child 2
  });
});
