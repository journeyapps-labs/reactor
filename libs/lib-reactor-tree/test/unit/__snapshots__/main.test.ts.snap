// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tree > should test all tree things 1`] = `
{
  "[[\\"child\\",\\"root\\"],[\\"child\\",\\"child1\\"],[\\"child\\",\\"child2\\"]]": {
    "collapsed": true,
  },
  "[[\\"child\\",\\"root\\"],[\\"child\\",\\"child1\\"]]": {
    "collapsed": true,
  },
  "[[\\"child\\",\\"root\\"]]": {
    "collapsed": true,
  },
}
`;
